ls1intum/Artemis
# Add course group management

### Checklist

### Motivation and Context
This PR makes the course group management easier for instructors and admins. It allows them to click on the Access Groups (Students, Tutors, Instructors) of a course in the Course Administration in order to inspect them. This opens up the new course group page that displays all users that belong to the selected course group in a table layout.

Two different actions can be performed on the page:
- Users can be removed from the course group (requires confirmation in modal)
- Users can be added to the course group via the autocomplete search bar

### Description
1. Add missing server endpoints for querying users
2. Mirror new server endpoints into client services
3. Create new component `CourseGroupComponent`
  3.1 Based on generic data table `DataTableComponent` (our wrapper around ngx-datatable)
  3.2 Adds server side search for users on top
  3.3 Customizes the behavior for when a user is selected via autocomplete (selecting a non-member user will add them to the group; member and non-member differentiation via colors and icons)
  3.4 "Remove" button for each user to remove them from the course group (uses our delete dialog button with built in confirmation dialog)
4. Enhancements to `DataTableComponent` needed for `CourseGroupComponent` (but with high reuse potential)
  4.1 Expose autocomplete suggestion buttons for out of component UI customization
  4.2 Enable customizing the search and autocomplete select behavior
  4.3 Add state information for the server side search use case (loading, no results, failed, query too short)

### Steps for Testing


1. Log in to Artemis (as instructor or admin)
2. Navigate to Course Administration
3. Click on one of the Access Groups for a Course
4. Check that all Course Group members are loaded
5. Test removing a user from the Course Group
6. Search for a user that is already part of the group and verify that he is found
7. Search for a user that is not yet part of the group and add them to the group

### Screenshots
1. The new Course Group page can be accessed by clicking on these group names (for instructors of the respective course and for admins):
![OnPaste 20200326-013111](https://user-images.githubusercontent.com/7109225/77598698-c9886b80-6f02-11ea-8308-967c3935eeb8.png)

2. The new Course Group page showing all group members
![OnPaste 20200326-013211](https://user-images.githubusercontent.com/7109225/77598700-cab99880-6f02-11ea-8e26-2696e000152e.png)

3. Upon clicking on the "Remove" button, the modal explains what will happen and asks for confirmation
![OnPaste 20200326-013232](https://user-images.githubusercontent.com/7109225/77598701-cab99880-6f02-11ea-9cf7-aacc70b9fd08.png)

4. When searching for a user (by login or name) to add to the course group, a list of users is presented in an autocomplete. The ones with less opacity, a light green background and a group icon are already part of the course group (clicking on them is still possible but doesn't do anything apart from filtering for them). The regular options with white background, strong green hover color and "add user" icon are users that are not yet part of the course group (clicking on them will add them to the course group; this is indicated by a flashing green row background once added).
![OnPaste 20200326-013323](https://user-images.githubusercontent.com/7109225/77598702-cb522f00-6f02-11ea-8fde-5804afacd019.png)

5. When the search term is too short for a useful server-side search (since it would return either very many results or a too unspecific selection), an info badge informs the user that he needs to enter at least 3 characters (this nice little add-on is part of the data table itself and is now also available for Exercise Scores and Participations):
![OnPaste 20200326-014833](https://user-images.githubusercontent.com/7109225/77599182-35b79f00-6f04-11ea-9d4e-6acf3fa40725.png)

6. When the server search for users returned no results:
![OnPaste 20200326-013515](https://user-images.githubusercontent.com/7109225/77598705-cb522f00-6f02-11ea-9cd8-c743ddaa836b.png)

7. When the server search for users did not execute as expected (e.g. due to no internet):
![OnPaste 20200326-013545](https://user-images.githubusercontent.com/7109225/77598706-cbeac580-6f02-11ea-90a0-08dfc03e0c91.png)
